src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -92,6 +93,57 @@ public TextDocumentIdentifier GetTextDocumentIdentifier(FoldingRangeParams reque
         return foldingRanges;
     }
 
+    private class RazorFileFoldingRangeSyntaxWalker : SyntaxWalker
+    {
+        private readonly RazorSourceDocument _source;
+        internal List<FoldingRange> Ranges { get; }
+
+        public RazorFileFoldingRangeSyntaxWalker(RazorSourceDocument source)
+        {
+            _source = source;
+            Ranges = new List<FoldingRange>();
+        }
+
+        public override void VisitRazorDirective(RazorDirectiveSyntax node)
+        {
+            // first rule: we can fold consecutive usings
+            // if we have an existing range that ends the line *before* this current using directive or begins the line *after*,
+            // we extend the range one line
+            if (node.Body is RazorDirectiveBodySyntax syntax && string.Equals(syntax.Keyword.GetFirstToken().Content, "using"))
```

Comment:
Would be good to extract the logic from https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorFormattingPass.cs#L287 into a static helper method somewhere, so we can start to build up our library of code that deals with the syntax tree.

Reply:
I had no idea that existed. Thank you for the catch